extras/modem-modeswitch/modem-modeswitch.8
.TH MODEM_MODESWITCH 8 "November 2005" "" "Linux Administrator's Manual"
.SH NAME
modem-modeswitch \- udev callout to switch mobile broadband devices to their
intended modem mode
.SH SYNOPSIS
.BI modem-modeswitch
\fI--vendor <vendor id>\fP \fI--product <product id>\fP [\fI--debug\fP]
\fI--type <type>\fP \fI<devpath>\fP
.SH "DESCRIPTION"
.B modem-modeswitch
is normally called from a udev rule, to eject the driver CD that many mobile
broadband devices mount by default, and switch them into modem mode.
.SH USAGE
.B modem-modeswitch
switches the device into modem mode
.SH OPTIONS
The following commandline switches are supported:
.TP
.BI \-\-vendor\ <vid>
the USB vendor ID of the mobile broadband device to switch
.TP
.BI \-\-product\ <pid>
the USB product ID of the mobile broadband device to switch
.TP
.BI \-\-type\ <type>
the type of switch to perform (one of: option-zerocd, mobile-action-8280p)
.TP
.BI \-\-log\ <file>
log verbose debugging information about the switching process
.TP
.BI \-\-debug
print verbose debugging information about the switching process
.RE
.SH SEE ALSO
.BR udev (7)
.SH AUTHORS
Developed by Dan Williams <dcbw@redhat.com> based off code by Peter Henn
<p.henn@option.com>.



NEWS
udev 144
========

udev 143
========
Bugfixes.

The configure options have changed because another library needs to be
installed in a different location. Instead of exec_prefix and udev_prefix,
libdir, rootlibdir and libexecdir are used. The Details are explained in
the README file.

Event processes now get re-used after they handled an event. This reduces
the number of forks and the pressure on the CPU significantly, because
cloned event processes no longer cause page faults in the main daemon.
After the events have settled, a few worker processes stay around for
future events, all others get cleaned up.

To be able to use signalfd(), udev depends on kernel version 2.6.25 now.
Also inotify support is mandatory now to run udev.

The format of the queue exported by the udev damon has changed. There is
no longer a /dev/.udev/queue/ directory. The current event queue can be
accessed with udevadm settle and libudedv.

Libudev does not have the unstable API header anymore. From now on,
incompatible changes will be handled by bumping the library major version.

To build udev from the git tree gtk-doc is needed now. The tarballs will
build without it and contain the pre-built documentation. An online copy
is available here:
  http://www.kernel.org/pub/linux/utils/kernel/hotplug/libudev/

The tools from the udev-extras repository have been merged into the main
udev repository. Some of the extras have larger external dependencies, and
they can be disabled with the configure switch --disable-extras.

udev 142
========
Bugfixes.

The program vol_id and the library libvolume_id are removed from the
repository. Libvolume_id is merged with libblkid from the util-linux-ng
package. Persistent disk links for label and uuid depend on the
util-linux-ng version (2.15) of blkid now. Older versions of blkid
can not be used with udev.

Libudev allows to subscribe to udev events. To prevent unwanted messages
to be delivered, and waking up the subscribing process, a filter can be
installed, to drop messages inside a kernel socket filter. The filters
match on the <subsytem>:<devtype> properties of the device.
    This is part of the ongoing effort to replace HAL, and switch current
users over to directly use libudev.
    Libudev is still marked as experimental, and its interface might
eventually change if needed, but no major changes of the currently exported
interface are expected anymore, and a first stable release should happen
soon.

A too old kernel (2.6.21) or a kernel with CONFIG_SYSFS_DEPRECATED
is not supported since while and udevd will log an error message at
startup. It should still be able to boot-up, but advanced rules and system
services which depend on the information not available in the old sysfs
format will fail to work correctly.

DVB device naming is supplied by the kernel now. In case older kernels
need to be supported, the old shell script should be added to a compat
rules file.

udev 141
========
Bugfixes.

The processed udev events get send back to the netlink socket. Libudev
provides access to these events. This is work-in-progress, to replace
the DeviceKit daemon functionality directly with libudev. There are
upcoming kernel changes to allow non-root users to subcribe to these
events.

udev 140
========
Bugfixes.

"udevadm settle" now optionally accepts a range of events to wait for,
instead of waiting for "all" events.

udev 139
========
Bugfixes.

The installed watch for block device metadata changes is now removed
during event hadling, because some (broken) tools may be called from udev
rules and (wrongly) open the device with write access. After the finished
event handling the watch is restored.

udev 138
========
Bugfixes.

Device nodes can be watched for changes with inotify with OPTIONS="watch".
If closed after being opened for writing, a "change" uevent will occur.
/dev/disk/by-{label,uuid}/* symlinks will be automatically updated.

udev 137
========
Bugfixes.

The udevadm test command has no longer a --force option, nodes and symlinks
are always updated with a test run now.

The udevd daemon can be started with --resolve-names=never to avoid all user
and group lookups (e.g. in cut-down systems) or --resolve-names=late to
lookup user and groups every time events are handled.

udev 136
========
Bugfixes.

We are currently merging the Ubuntu rules in the udev default rules,
and get one step closer to provide a common Linux /dev setup, regarding
device names, symlinks, and default device permissions. On udev startup,
we now expect the following groups to be resolvable to their ids with
glibc's getgrnam():
  disk, cdrom, floppy, tape, audio, video, lp, tty, dialout, kmem.
LDAP setups need to make sure, that these groups are always resolvable at
bootup, with only the rootfs mounted, and without network access available.

Some systems may need to add some new, currently not used groups, or need
to add some users to new groups, but the cost of this change is minimal,
compared to the pain the current, rather random, differences between the
various distributions cause for upstream projects and third-party vendors.

In general, "normal" users who log into a machine should never be a member
of any such group, but the device-access should be managed by dynamic ACLs,
which get added and removed for the specific users on login/logout and
session activity/inactivity. These groups are only provided for custom setups,
and mainly system services, to allow proper privilege separation.
A video-streaming daemon uid would be a member of "audio" and "video", to get
access to the sound and video devices, but no "normal" user should ever belong
to the "audio" group, because he could listen to the built-in microphone with
any ssh-session established from the other side of the world.

/dev/serial/by-{id,path}/ now contains links for ttyUSB devices,
which do not depend on the kernel device name. As usual, unique
devices - only a single one per product connected, or a real

extras/modem-modeswitch/modem-modeswitch.c
/*
 * Modem mode switcher
 *
 * Copyright (C) 2008  Dan Williams <dcbw@redhat.com>
 * Copyright (C) 2008  Peter Henn <support@option.com>
 *
 * Heavily based on the 'ozerocdoff' tool by Peter Henn.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details:
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <stdarg.h>
#include <getopt.h>

#include <usb.h>

#include "utils.h"

#include "ma8280p_us.h"
#include "option.h"

struct usb_dev_handle *handle = NULL;

typedef struct usb_device * (*FindFunc) (int vid, int pid);
typedef int (*SwitchFunc) (struct usb_dev_handle *dh, struct usb_device *dev);

typedef enum {
	ST_UNKNOWN = 0,
	ST_OPTION_ZEROCD,
	ST_MA8280P
} SwitchType;

typedef struct SwitchEntry {
	SwitchType st;
	const char *clopt;
	FindFunc find_func;
	SwitchFunc switch_func;
} SwitchEntry;

static SwitchEntry switch_types[] = {
	{ ST_OPTION_ZEROCD, "option-zerocd", option_zerocd_find, option_zerocd_switch },
	{ ST_MA8280P, "mobile-action-8280p", NULL, ma8280p_switch },
	{ ST_UNKNOWN, NULL, NULL }
};

static struct usb_device *
generic_find (int vid, int pid)
{
	struct usb_bus *bus;
	struct usb_device *dev;

	for (bus = usb_get_busses(); bus; bus = bus->next) {
		for (dev = bus->devices; dev; dev = dev->next) {
			if (dev->descriptor.idVendor == vid && dev->descriptor.idProduct == pid) {
				debug ("Found device '%s'", dev->filename);
				return dev;
			}
		}
	}
	return NULL;
}

static void
release_usb_device (int param)
{
	usb_release_interface (handle, 0);
	usb_close (handle);
}

static void
print_usage (void)
{
	printf ("Usage: modem-modeswitch [-hdq] [-l <file>] -v <vendor-id> -p <product-id> -t <type>\n"
	        " -h, --help               show this help message\n"
	        " -v, --vendor <n>         target USB vendor ID\n"
	        " -p, --product <n>        target USB product ID\n"
	        " -t, --type <type>        type of switch to attempt, varies by device:\n"
	        "                               option-zerocd         - For many newer Option N.V. devices\n"
	        "                               mobile-action-8280p   - For Mobile Action 8xxxP USB cables\n"
	        " -l, --log <file>         log output to a file\n"
	        " -q, --quiet              don't print anything to stdout\n"
	        " -d, --debug              display debugging messages\n\n"
	        "Examples:\n"
	        "   modem-modeswitch -v 0x0af0 -p 0xc031 -t option-zerocd\n");
}

static SwitchEntry *
parse_type (const char *s)
{
	SwitchEntry *entry = &switch_types[0];

	while (entry->clopt) {
		if (!strcmp (entry->clopt, s))
			return entry;
		entry++;
	}

	return NULL;
}

static void
do_exit (int val)
{
	log_shutdown ();
	exit (val);
}

int main(int argc, char **argv)
{
	static struct option options[] = {
		{ "help",	 no_argument,       NULL, 'h' },
		{ "vendor",  required_argument, NULL, 'v' },
		{ "product", required_argument, NULL, 'p' },
		{ "type",    required_argument, NULL, 't' },
		{ "log",     required_argument, NULL, 'l' },
		{ "debug",   no_argument,       NULL, 'd' },
		{ "quiet",   no_argument,       NULL, 'q' },
		{ NULL, 0, NULL, 0}
	};

	struct usb_device *dev;
	int vid = 0, pid = 0;
	const char *logpath = NULL;
	char buffer[256];
	int ret, quiet = 0, debug = 0;
	SwitchEntry *sentry = NULL;

	while (1) {
		int option;

		option = getopt_long(argc, argv, "hv:p:l:t:dq", options, NULL);
		if (option == -1)
			break;

		switch (option) {
		case 'v':
			vid = strtol (optarg, NULL, 0);
			break;
		case 'p':
			pid = strtol (optarg, NULL, 0);
			break;
		case 't':
			sentry = parse_type (optarg);
			if (!sentry) {
				error ("unknown switch type '%s'", optarg);
				print_usage ();
				exit (1);
			}
			break;
		case 'l':
			logpath = optarg;
			break;
		case 'q':
			quiet = 1;
			break;
		case 'd':
			debug = 1;
			break;
		case 'h':
		default:
			print_usage ();
			exit (1);
		}
	}

	if (log_startup (logpath, debug, quiet)) {
		fprintf (stderr, "Couldn't open/create logfile %s", logpath);
		exit (2);
	}

	if (!sentry) {
		if (!quiet)
			print_usage ();
		else
			error ("missing device switch type.");
		do_exit (3);
	}

	if (!vid || !pid) {
		if (!quiet)
			print_usage ();
		else
			error ("missing vendor and device IDs.");
		do_exit (3);
	}

	usb_init();

	if (usb_find_busses() < 0) {
		error ("no USB busses found.");
		do_exit (4);
	}

	if (usb_find_devices() < 0) {
		error ("no USB devices found.");
		do_exit (4);
	}

	if (sentry->find_func)
		dev = (*sentry->find_func) (vid, pid);
	else
		dev = generic_find (vid, pid);
	if (dev == NULL) {
		error ("no device found.");
		do_exit (5);
	}

	handle = usb_open (dev);
	if (handle == NULL) {
		error ("%s: could not access the device.",
		         dev->filename);
		do_exit (6);
	}

	/* detach running default driver */
	signal (SIGTERM, release_usb_device);
	ret = usb_get_driver_np (handle, 0, buffer, sizeof (buffer));
	if (ret == 0) {
		debug ("%s: found already attached driver '%s'", dev->filename, buffer);

		ret = usb_detach_kernel_driver_np (handle, 0);
		if (ret != 0) {
			debug ("%s: error: unable to detach current driver.", dev->filename);
			usb_close (handle);
			do_exit (7);
		}
	}

	ret = usb_claim_interface (handle, 0);
	if (ret != 0) {
		debug ("%s: couldn't claim device's USB interface: %d.",
		       dev->filename, ret);
		usb_close (handle);
		do_exit (8);
	}

	ret = (*sentry->switch_func) (handle, dev);
	if (ret < 0) {
		debug ("%s: failed to switch device to modem mode.", dev->filename);
		usb_release_interface (handle, 0);
		usb_close (handle);
		do_exit(9);
	}

	usb_release_interface (handle, 0);

	ret = usb_close (handle);
	if (ret < 0)
		debug ("%s: failed to close the device.", dev->filename);

	do_exit (0);
	return 0;
}
